go/types/graph_builder.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/util/status"
+	"github.com/dustin/go-humanize"
+)
+
+type GraphBuilder struct {
+	oc      opCache
+	vrw     ValueReadWriter
+	stack   graphStack
+	verbose bool
+}
+
+// NewGraphBuilder() returns an object that is used in building a graph of
+// objects that are stored on disk and then assembled into a noms collection.
+// Graphs can exist of nested maps, with maps, sets and lists at the outermost
+// levels of the graph. Any of the collections (nested or leaf) can contain onto
+// any type of Value. |rootKind| indicates what the topLevel collection should be.
+// It must be one of: MapKind, SetKind, or ListKind. If this GraphBuilder is
+// intended to build nested maps, |rootKind| must be MapKind.
+func NewGraphBuilder(vrw ValueReadWriter, rootKind NomsKind, verbose bool) *GraphBuilder {
+	return newGraphBuilder(vrw, vrw.opCache(), rootKind, verbose)
+}
+
+func newGraphBuilder(vrw ValueReadWriter, opc opCache, rootKind NomsKind, verbose bool) *GraphBuilder {
+	b := &GraphBuilder{oc: opc, vrw: vrw, verbose: verbose}
+	b.pushNewKeyOnStack(String("ROOT"), rootKind, nil)
+	return b
+}
+
+// MapSet() will add the key/value pair |k, v| to the map found by traversing
+// the graph using the |keys| slice. Intermediate maps referenced by |keys| are
+// created as necessary
+func (b *GraphBuilder) MapSet(keys []Value, k Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call MapSet() again after Build()")
+	b.oc.GraphMapSet(keys, k, v)
+}
+
+// SetInsert() will insert the value |v| into the set at path |keys|. Intermediate
+// maps referenced by |keys| are created as necessary.
+func (b *GraphBuilder) SetInsert(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call SetInsert() again after Build()")
+	b.oc.GraphSetInsert(keys, v)
+}
+
+// ListAppends() will append |v| to the list at path |p|. Intermediate
+// maps referenced by |keys| are created as necessary.
+func (b *GraphBuilder) ListAppend(keys []Value, v Value) {
+	d.Chk.True(b.oc != nil, "Can't call ListAppend() again after Build()")
+	b.oc.GraphListAppend(keys, v)
+}
+
+type GraphOpContainer struct {
+	keys []Value
+	kind NomsKind
+	item sequenceItem
+}
+
+// Builds and returns the graph.
+func (b *GraphBuilder) Build() Value {
+	d.Chk.True(b.oc != nil, "Can only call Finish() once")
```

Comment:
I talked to Raf about this yesterday and his issue I think was that this could be called once by two different threads and possibly not have this check fail since oc didn't get set to nil until the deferred function was called. I can fix that pretty easily by coping the oc to a local variable and setting oc to nil immediately.

I did think of another issue about calling the mutation functions from multiple threads though. The order that values get appended to lists depends on the order that ListAppend() is called. So any one who cared about that would have to manage when that method got called.